src/Razor/src/Microsoft.AspNetCore.Razor.Language/RazorHtmlLoweringPass.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Text;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal class RazorHtmlLoweringPass : IntermediateNodePassBase, IRazorDocumentClassifierPass
+    {
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
```

Comment:
lol totally would have done the exact same thing if I wasn't handed off an issue earlier today about Blazor perf. Would it be possible to lazily evaluate the HTML document in `codeDocument.GetHtmlDocument` and memoize it instead of having a pass? That way we can avoid creating an HTML document with large strings if we don't need to.